app/src/main/java/com/android/unio/ui/components/EventEditComponents.kt

```diff
@@ -0,0 +1,317 @@
+                val upEvent = waitForUpOrCancellation(pass = PointerEventPass.Initial)
+                if (upEvent != null) {
+                  isTimePickerVisible = true
+                }
+              }
+            },
+        value =
+            selectedTime?.let { convertMillisToTime(it) }
+                ?: initialTime?.let { convertMillisToTime(it) }
+                ?: "",
+        readOnly = true,
+        onValueChange = {},
+        trailingIcon = { Icon(Icons.Default.AccessTime, contentDescription = "Select date") },
+        placeholder = { Text(context.getString(R.string.event_creation_placeholder_time_input)) },
+        label = { Text(timeString) })
+  }
+
+  if (isDatePickerVisible) {
+    DatePickerModal(
+        modifier = Modifier.testTag(datePickerTestTag),
+        onDateSelected = {
+          selectedDate = it
+          isDatePickerVisible = false
+        },
+        onDismiss = { isDatePickerVisible = false })
+  }
+
+  if (isTimePickerVisible) {
+    TimePickerModal(
+        modifier = Modifier.testTag(timePickerTestTag),
+        onTimeSelected = {
+          selectedTime = it
+          isTimePickerVisible = false
+        },
+        onDismiss = { isTimePickerVisible = false })
+  }
+
+  // Allows to only partially fill the date and time fields
+  if (selectedDate != null && selectedTime != null) {
+    onTimestamp(Timestamp(Date(selectedDate!! + selectedTime!!)))
+  } else if (selectedDate != null && initialTime != null) {
+    onTimestamp(Timestamp(Date(selectedDate!! + initialTime)))
+  } else if (initialDate != null && selectedTime != null) {
+    onTimestamp(Timestamp(Date(initialDate + selectedTime!!)))
+  }
+}
+
+@OptIn(ExperimentalMaterial3Api::class)
+@Composable
+fun DatePickerModal(
+    onDateSelected: (Long?) -> Unit,
+    onDismiss: () -> Unit,
+    modifier: Modifier = Modifier
+) {
+  val datePickerState = rememberDatePickerState()
+  val context = LocalContext.current
+
+  DatePickerDialog(
+      modifier = modifier,
+      onDismissRequest = onDismiss,
+      confirmButton = {
+        TextButton(
+            onClick = {
+              onDateSelected(datePickerState.selectedDateMillis)
+              onDismiss()
+            }) {
+              Text(context.getString(R.string.event_creation_dialog_ok))
+            }
+      },
+      dismissButton = {
+        TextButton(onClick = onDismiss) {
+          Text(context.getString(R.string.event_creation_dialog_cancel))
+        }
+      }) {
+        DatePicker(state = datePickerState)
+      }
+}
+
+@OptIn(ExperimentalMaterial3Api::class)
+@Composable
+fun TimePickerModal(onTimeSelected: (Long?) -> Unit, onDismiss: () -> Unit, modifier: Modifier) {
+  val timePickerState = rememberTimePickerState(is24Hour = true)
+
+  TimePickerDialog(
+      modifier = modifier,
+      onDismiss = onDismiss,
+      onConfirm = {
+        onTimeSelected(
+            timePickerState.hour * 60 * 60 * 1000L - 3600000 + timePickerState.minute * 60 * 1000L)
```

Comment:
Consider extracting this in a separate method to make the code more readable, just like you did with the other utils methods.